File: apps/desktop/src/services/commands/backup.ts
import { invoke } from "@tauri-apps/api/core";

export interface CreateBackupInputDto {
  filePath?: string | null;
  note?: string | null;
  includeContent?: boolean;
  overwrite?: boolean;
}

export interface RestoreBackupInputDto {
  filePath: string;
  confirmRestore: boolean;
  createSafetyBackup?: boolean;
}

export interface BackupTableCountDto {
  table: string;
  rows: number;
}

export interface BackupExportResultDto {
  backupId: number;
  filePath: string;
  bytesWritten: number;
  createdAt: string;
  backupKind: string;
  includeContent: boolean;
  tableCounts: BackupTableCountDto[];
}

export interface BackupHistoryItemDto {
  id: number;
  filePath: string;
  fileSizeBytes: number | null;
  note: string | null;
  backupKind: string;
  includeContent: boolean;
  createdAt: string;
  exists: boolean;
}

export interface BackupHistoryDto {
  items: BackupHistoryItemDto[];
  total: number;
  latestBackupAt: string | null;
  latestAutoBackupAt: string | null;
  backupDirectory: string;
}

export interface BackupValidationDto {
  valid: boolean;
  schema: string | null;
  schemaVersion: number | null;
  exportedAt: string | null;
  backupKind: string | null;
  username: string | null;
  includeContent: boolean;
  tableCounts: BackupTableCountDto[];
  warnings: string[];
}

export interface BackupRestoreResultDto {
  restoredAt: string;
  restoredTables: BackupTableCountDto[];
  safetyBackup: BackupHistoryItemDto | null;
}

export interface ScheduledBackupResultDto {
  created: boolean;
  skippedReason: string | null;
  backup: BackupExportResultDto | null;
}

export function createBackup(
  input: CreateBackupInputDto,
): Promise<BackupExportResultDto> {
  return invoke<BackupExportResultDto>("create_backup", { input });
}

export function listBackups(): Promise<BackupHistoryDto> {
  return invoke<BackupHistoryDto>("list_backups");
}

export function validateBackup(filePath: string): Promise<BackupValidationDto> {
  return invoke<BackupValidationDto>("validate_backup", { filePath });
}

export function restoreBackup(
  input: RestoreBackupInputDto,
): Promise<BackupRestoreResultDto> {
  return invoke<BackupRestoreResultDto>("restore_backup", { input });
}

export function ensureScheduledBackup(): Promise<ScheduledBackupResultDto> {
  return invoke<ScheduledBackupResultDto>("ensure_scheduled_backup");
}

File: apps/desktop/src-tauri/src/dto/backup.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateBackupInputDto {
    pub file_path: Option<String>,
    pub note: Option<String>,
    pub include_content: Option<bool>,
    pub overwrite: Option<bool>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RestoreBackupInputDto {
    pub file_path: String,
    pub confirm_restore: bool,
    pub create_safety_backup: Option<bool>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BackupTableCountDto {
    pub table: String,
    pub rows: usize,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BackupExportResultDto {
    pub backup_id: i64,
    pub file_path: String,
    pub bytes_written: u64,
    pub created_at: String,
    pub backup_kind: String,
    pub include_content: bool,
    pub table_counts: Vec<BackupTableCountDto>,
}

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct BackupHistoryItemDto {
    pub id: i64,
    pub file_path: String,
    pub file_size_bytes: Option<i64>,
    pub note: Option<String>,
    pub backup_kind: String,
    pub include_content: bool,
    pub created_at: String,
    pub exists: bool,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BackupHistoryDto {
    pub items: Vec<BackupHistoryItemDto>,
    pub total: usize,
    pub latest_backup_at: Option<String>,
    pub latest_auto_backup_at: Option<String>,
    pub backup_directory: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BackupValidationDto {
    pub valid: bool,
    pub schema: Option<String>,
    pub schema_version: Option<u32>,
    pub exported_at: Option<String>,
    pub backup_kind: Option<String>,
    pub username: Option<String>,
    pub include_content: bool,
    pub table_counts: Vec<BackupTableCountDto>,
    pub warnings: Vec<String>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BackupRestoreResultDto {
    pub restored_at: String,
    pub restored_tables: Vec<BackupTableCountDto>,
    pub safety_backup: Option<BackupHistoryItemDto>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ScheduledBackupResultDto {
    pub created: bool,
    pub skipped_reason: Option<String>,
    pub backup: Option<BackupExportResultDto>,
}
